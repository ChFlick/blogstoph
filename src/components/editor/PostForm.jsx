import React from 'react';
import moment from 'moment';
import Markdown from 'react-remarkable';
import { SingleDatePicker } from 'react-dates';

class PostForm extends React.Component {
    constructor(props) {
        super(props);

        const post = props.post || {};
        this.state = {
            title: post.title || '',
            content: post.content || '',
            author: post.author || '',
            date: moment(post.date) || moment(),
            dateFocused: false,
            published: post.published || false,
            error: ''
        };
    }

    onSubmit = (e) => {
        e.preventDefault();

        const { title, content, author, date, published } = this.state;

        if (published && (!title || !content || !author || !date)) {
            this.setState(() => ({
                error: 'Please provide all information when publishing a post.'
            }));
        } else if (!title) {
            this.setState(() => ({
                error: 'Please provide a title to identify the post.'
            }));
        } else {
            this.setState(() => ({
                error: ''
            }));

            this.props.onSubmit({
                title,
                content,
                author,
                date: date.valueOf(),
                published
            });
        }
    };

    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({
            title
        }));
    };

    onContentChange = (e) => {
        const content = e.target.value;
        this.setState(() => ({
            content
        }));
    };

    onAuthorChange = (e) => {
        const author = e.target.value;
        this.setState(() => ({
            author
        }));
    };

    onDateChange = (date) => {
        this.setState(() => ({ date }));
    };

    onPublishedChange = (e) => {
        const published = e.target.checked;
        this.setState(() => ({
            published
        }));
    };

    createTitleInput = () => (
        <input
            type="text"
            placeholder="Title *"
            onChange={this.onTitleChange}
            value={this.state.title}
            className="text-input"
        />
    );

    createContentInput = () => (
        <textarea
            type="text"
            placeholder="Content"
            onChange={this.onContentChange}
            value={this.state.content}
            className="textarea"
        />
    );

    createAuthorInput = () => (
        <input
            type="text"
            placeholder="Author"
            onChange={this.onAuthorChange}
            value={this.state.author}
            className="text-input"
        />
    );

    createDatePicker = () => (
        <SingleDatePicker
            date={this.state.date}
            onDateChange={this.onDateChange}
            focused={this.state.dateFocused}
            onFocusChange={({ focused }) => this.setState({ dateFocused: focused })}
            numberOfMonths={1}
            displayFormat="DD.MM.YYYY"
        />
    );

    createPublishedCheckbox = () => (
        <div>
            <div className="pretty p-svg p-curve p-smooth">
                <input
                    type="checkbox"
                    onChange={this.onPublishedChange}
                    checked={this.state.published}
                />
                <div className="state p-success">
                    <svg className="svg svg-icon" viewBox="0 0 20 20">
                        <path d="M7.629,14.566c0.125,0.125,0.291,0.188,0.456,0.188c0.164,0,
                        0.329-0.062,0.456-0.188l8.219-8.221c0.252-0.252,
                        0.252-0.659,0-0.911c-0.252-0.252-0.659-0.252-0.911,
                        0l-7.764,7.763L4.152,9.267c-0.252-0.251-0.66-0.251-0.911,
                        0c-0.252,0.252-0.252,0.66,0,0.911L7.629,14.566z" style={{stroke: 'white', fill: 'white'}}></path>
                    </svg>
                    <label>Published</label>
                </div>
            </div>
        </div>
    );

    createButtons = () => (
        <div>
            <button
                className="button button--ingroup"
                type="button"
                onClick={this.props.onBack}
            >
                Back
            </button>
            <button className="button button--ingroup">
                Save Post
            </button>
        </div>
    );

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                {this.createTitleInput()}
                {this.createContentInput()}
                <Markdown source={this.state.content} />
                {this.createAuthorInput()}
                {this.createDatePicker()}
                {this.createPublishedCheckbox()}
                {this.createButtons()}
            </form>
        );
    };
}

export default PostForm;
