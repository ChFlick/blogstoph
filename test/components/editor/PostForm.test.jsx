import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

import PostForm from '../../../src/components/editor/PostForm';

const testText = 'TEST';

test('renders the post form correctly', () => {
    const wrapper = shallow(<PostForm />);

    expect(wrapper).toMatchSnapshot();
});

describe('the post form', () => {
    test('should set the title', () => {
        const wrapper = shallow(<PostForm />);

        wrapper.find('#title').simulate('change', { target: { value: testText } });
        expect(wrapper.state('title')).toEqual(testText);
    });

    test('should set the subtitle', () => {
        const wrapper = shallow(<PostForm />);

        wrapper.find('#subtitle').simulate('change', { target: { value: testText } });
        expect(wrapper.state('subtitle')).toEqual(testText);
    });


    test('should set the content', () => {
        const wrapper = shallow(<PostForm />);

        wrapper.find('#content').simulate('change', { target: { value: testText } });
        expect(wrapper.state('content')).toEqual(testText);
    });

    test('should set the author', () => {
        const wrapper = shallow(<PostForm />);

        wrapper.find('#author').simulate('change', { target: { value: testText } });
        expect(wrapper.state('author')).toEqual(testText);
    });

    test('should set the date', () => {
        const newDate = moment(0);
        const wrapper = shallow(<PostForm />);

        wrapper.find(SingleDatePicker).prop('onDateChange')(newDate);
        expect(wrapper.state('date')).toEqual(newDate);
    });

    test('should set published flag', () => {
        const published = true;
        const wrapper = shallow(<PostForm />);

        wrapper.find('#published').simulate('change', { target: { checked: published } });
        expect(wrapper.state('published')).toEqual(published);
    });
});


test('allows to set a post from outside', () => {
    const post = {
        title: 'test',
        author: 'hello',
        date: 0,
        published: true
    };

    const wrapper = shallow(<PostForm post={post} />);

    expect(wrapper.state('title')).toEqual(post.title);
    expect(wrapper.state('author')).toEqual(post.author);
    expect(wrapper.state('date').valueOf()).toEqual(post.date);
    expect(wrapper.state('published')).toEqual(post.published);
    expect(wrapper.state('content')).toEqual('');
});

describe('in onSubmit', () => {
    test('shouldn\'t call onSubmit at submission with no data', () => {
        const onSubmitSpy = jest.fn();
        const wrapper = shallow(<PostForm onSubmit={onSubmitSpy} />);

        wrapper.find('form').simulate('submit', { preventDefault: () => { } });

        expect(onSubmitSpy).not.toHaveBeenCalled();
    });

    test('should set error without data even when private', () => {
        const wrapper = shallow(<PostForm onSubmit={jest.fn()} />);

        wrapper.find('form').simulate('submit', { preventDefault: () => { } });

        expect(wrapper.state('error')).toEqual(expect.any(String));
    });

    test('shouldn\'t set error with only title when private', () => {
        const wrapper = shallow(<PostForm onSubmit={jest.fn()} />);
        wrapper.setState({ title: 'test' });

        wrapper.find('form').simulate('submit', { preventDefault: () => { } });

        expect(wrapper.state('error')).toEqual('');
    });

    test('should call onSubmit with only title when private', () => {
        const onSubmitSpy = jest.fn();
        const wrapper = shallow(<PostForm onSubmit={onSubmitSpy} />);
        wrapper.setState({ title: 'test' });

        wrapper.find('form').simulate('submit', { preventDefault: () => { } });

        expect(onSubmitSpy).toHaveBeenCalled();
    });

    test('should set error with only title when public', () => {
        const wrapper = shallow(<PostForm onSubmit={jest.fn()} />);
        wrapper.setState({ title: 'test', published: true });
        wrapper.find('form').simulate('submit', { preventDefault: () => { } });

        expect(wrapper.state('error')).toEqual(expect.any(String));
    });

    test('shouldn\'t set error with data when public', () => {
        const wrapper = shallow(<PostForm onSubmit={jest.fn()} />);

        wrapper.setState({
            title: 'test',
            subtitle: 'test',
            content: 'test',
            author: 'someone',
            published: true
        });

        wrapper.find('form').simulate('submit', { preventDefault: () => { } });

        expect(wrapper.state('error')).toEqual('');
    });

    test('should call onSubmit with data', () => {
        const onSubmitSpy = jest.fn();
        const wrapper = shallow(<PostForm onSubmit={onSubmitSpy} />);

        wrapper.setState({
            title: 'test',
            subtitle: 'test',
            content: 'test',
            author: 'someone',
            published: true
        });

        wrapper.find('form').simulate('submit', { preventDefault: () => { } });

        expect(onSubmitSpy).toHaveBeenCalled();
    });
});
