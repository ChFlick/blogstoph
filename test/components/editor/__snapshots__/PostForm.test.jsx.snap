// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the post form correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    className="text-input"
    id="title"
    onChange={[Function]}
    placeholder="Title *"
    type="text"
    value=""
  />
  <input
    className="text-input"
    id="subtitle"
    onChange={[Function]}
    placeholder="Subtitle"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    id="content"
    onChange={[Function]}
    placeholder="Content"
    type="text"
    value=""
  />
  <ReactMarkdown
    astPlugins={Array []}
    className="post"
    escapeHtml={true}
    plugins={Array []}
    rawSourcePos={false}
    renderers={Object {}}
    skipHtml={false}
    source=""
    sourcePos={false}
    transformLinkUri={[Function]}
  />
  <input
    className="text-input"
    id="author"
    onChange={[Function]}
    placeholder="Author"
    type="text"
    value=""
  />
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    daySize={39}
    disabled={false}
    displayFormat="DD.MM.YYYY"
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    id="date"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonth={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
  <div>
    <div
      className="pretty p-svg p-curve p-smooth"
    >
      <input
        checked={false}
        id="published"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="state p-success"
      >
        <svg
          className="svg svg-icon"
          viewBox="0 0 20 20"
        >
          <path
            d="M7.629,14.566c0.125,0.125,0.291,0.188,0.456,0.188c0.164,0,
 0.329-0.062,0.456-0.188l8.219-8.221c0.252-0.252,
 0.252-0.659,0-0.911c-0.252-0.252-0.659-0.252-0.911,
 0l-7.764,7.763L4.152,9.267c-0.252-0.251-0.66-0.251-0.911,
 0c-0.252,0.252-0.252,0.66,0,0.911L7.629,14.566z"
            style={
              Object {
                "fill": "white",
                "stroke": "white",
              }
            }
          />
        </svg>
        <label>
          Published
        </label>
      </div>
    </div>
  </div>
  <div>
    <button
      className="button button--ingroup"
      type="button"
    >
      Back
    </button>
    <button
      className="button button--ingroup"
    >
      Save Post
    </button>
  </div>
</form>
`;
